pkg/bot/discord.go

```diff
@@ -0,0 +1,154 @@
+	BotID         string
+	Request       string
+	Response      string
+	IsAuthChannel bool
+	Session       *discordgo.Session
+}
+
+// NewDiscordBot returns new Bot object
+func NewDiscordBot(c *config.Config) Bot {
+	return &DiscordBot{
+		Token:            c.Communications.Discord.Token,
+		BotID:            c.Communications.Discord.BotID,
+		AllowKubectl:     c.Settings.Kubectl.Enabled,
+		RestrictAccess:   c.Settings.Kubectl.RestrictAccess,
+		ClusterName:      c.Settings.ClusterName,
+		ChannelID:        c.Communications.Discord.Channel,
+		DefaultNamespace: c.Settings.Kubectl.DefaultNamespace,
+	}
+}
+
+// Start starts the DiscordBot websocket connection and listens for messages
+func (b *DiscordBot) Start() {
+	api, err := discordgo.New("Bot " + b.Token)
+	if err != nil {
+		log.Error("error creating Discord session,", err)
+		return
+	}
+
+	// Register the messageCreate func as a callback for MessageCreate events.
+	api.AddHandler(func(s *discordgo.Session, m *discordgo.MessageCreate) {
+		dm := discordMessage{
+			Event:   m,
+			BotID:   b.BotID,
+			Session: s,
+		}
+
+		dm.HandleMessage(b)
+	})
+
+	// Open a websocket connection to Discord and begin listening.
+	go func() {
+		err := api.Open()
+		if err != nil {
+			log.Error("error opening connection,", err)
+			return
+		}
+	}()
+
+	log.Info("BotKube connected to Discord!")
+}
+
+// HandleMessage handles the incomming messages
+func (dm *discordMessage) HandleMessage(b *DiscordBot) {
+
+	// Serve only if starts with mention
+	if !strings.HasPrefix(dm.Event.Content, "<@!"+dm.BotID+"> ") && !strings.HasPrefix(dm.Event.Content, "<@"+dm.BotID+"> ") {
+		return
+	}
+
+	log.Debugf("Discord incoming message: %+v", dm.Event.Content)
+
+	// Serve only if current channel is in config
+	if b.ChannelID == dm.Event.ChannelID {
+		dm.IsAuthChannel = true
+	}
+
+	// Trim the @BotKube prefix
+	if strings.HasPrefix(dm.Event.Content, "<@!"+dm.BotID+"> ") {
+		dm.Request = strings.TrimPrefix(dm.Event.Content, "<@!"+dm.BotID+"> ")
+	} else if strings.HasPrefix(dm.Event.Content, "<@"+dm.BotID+"> ") {
+		dm.Request = strings.TrimPrefix(dm.Event.Content, "<@"+dm.BotID+"> ")
+	}
+
+	if len(dm.Request) == 0 {
+		return
+	}
+
+	e := execute.NewDefaultExecutor(dm.Request, b.AllowKubectl, b.RestrictAccess, b.DefaultNamespace,
+		b.ClusterName, config.DiscordBot, b.ChannelID, dm.IsAuthChannel)
+
+	dm.Response = e.Execute()
+	dm.Send()
+}
+
+func (dm discordMessage) Send() {
+	// Upload message as a file if too long
+	if len(dm.Response) >= 2000 {
+		params := &discordgo.MessageSend{
+			Content: dm.Request,
+			Files: []*discordgo.File{
+				{
+					Name:   "Response",
+					Reader: strings.NewReader(dm.Response),
+				},
+			},
+		}
+
+		if _, err := dm.Session.ChannelMessageSendComplex(dm.Event.ChannelID, params); err != nil {
+			log.Error("Error in uploading file:", err)
+		}
+		return
+	} else if len(dm.Response) == 0 {
+		log.Info("Invalid request. Dumping the response")
```

Comment:
We should log request message also. I know this is missing in other Bot interface implementations, but this will make things debug easier

Reply:
added a debug log for the same after INFO `log.Debugf("Dumped responce : %+v", dm.response)`

Reply:
added Info level to log request sent to botkube, and the corresponding valid responses generated by botkube.

Reply:
created a separate PR #404, for adding valid Request and Response log, and to remove capture of all events in slack.